### PR TITLE
Move number parser to stately, and refactor controlled selection behavior

### PR DESCRIPTION
**Pull request by devongovett**
This refactors number field to move the number parser to the new `@react-stately/i18n` package rather than React Aria. This allows it to be used directly by `@react-stately/numberfield` rather than needing to be passed in, which simplifies a lot of things.

In addition, the number parser has been refactored. It's now a class internally, which is cached based on the locale and format options. A `parseNumber` function is exported which takes a locale, options, and value to parse. It automatically detects the numbering format before parsing, and uses the matching number parser.

An `isValidPartialNumber` function is also exported from `@react-stately/i18n`, which returns whether a partial number string is valid according to the locale and detected numbering system. This replaces the previous `clean` function, and corresponds to a change to the way React Aria works. Rather than handling `onChange`, and trying to keep the selection correct when entering invalid characters, we use `onBeforeInput` to `preventDefault` on characters that are invalid. This is done by computing what the new value will be, and using the `isValidPartialNumber` function to determine whether it's valid. This also fixes cases where we were previously simply removing all non-numeric values, e.g. `a1b2c3` would end up as `123`. Now this is not accepted.

### CODE REVIEW

File: patches/@testing-library+user-event+12.1.3.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/node_modules/@testing-library/user-event/dist/type.js b/node_modules/@testing-library/user-event/dist/type.js
+index fd7d7b9..6494d98 100644
+--- a/node_modules/@testing-library/user-event/dist/type.js
++++ b/node_modules/@testing-library/user-event/dist/type.js
+@@ -304,6 +304,7 @@ function typeCharacter(char, {
+       key,
+       keyCode,
+       charCode: keyCode,
++      which: keyCode,
```
Comment (devongovett) [inline patches/@testing-library+user-event+12.1.3.patch:9]: This is a bug in user-event. I will submit a PR.

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 2]
```diff
@@ -54,92 +48,37 @@ export function useNumberFieldState(
...
-    if (!isNaN(value)) {
-      return value;
-    } else if (!isNaN(defaultValue)) {
-      return defaultValue;
-    }
-    return NaN;
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
-  let lastValDispatched = useRef(startingValue);
-  let smartOnChange = useCallback((val) => {
```
Comment (devongovett) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:91]: No longer necessary due to #1480

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 3]
```diff
@@ -54,92 +48,37 @@ export function useNumberFieldState(
...
-
-  // this removes any not allowed characters from the input value
-  let cleanInputValue = useMemo(() => numberParser.clean(inputValue), [numberParser, inputValue]);
-  // Number parser doesn't know about min/max, so we must remove it ourselves
-  if (minValue >= 0) {
-    cleanInputValue = cleanInputValue.replace(minusSign, '');
-  }
-  if (maxValue <= 0) {
-    cleanInputValue = cleanInputValue.replace(plusSign, '');
-  }
```
Comment (devongovett) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:136]: This is now handled by `isValidPartialNumber`

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 4]
```diff
@@ -54,92 +48,37 @@ export function useNumberFieldState(
...
-    setCurrentNumeralSystem
+    locale
   } = props;
 
-  let intlOptions = useMemo(() => inputValueFormatter.resolvedOptions(), [inputValueFormatter]);
+  let [numberValue, setNumberValue] = useControlledState<number>(value, isNaN(defaultValue) ? NaN : defaultValue, onChange);
+  let [inputValue, setInputValue] = useState(() => isNaN(numberValue) ? '' : getNumberFormatter(locale, formatOptions).format(numberValue));
+
+  let numberingSystem = useMemo(() => getNumberingSystem(inputValue), [inputValue]);
+  let formatter = useMemo(() => getNumberFormatter(locale, {...formatOptions, numberingSystem}), [locale, formatOptions, numberingSystem]);
```
Comment (snowystinger) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:58]: getNumberFormatter already has a cache, do we want or need to cache it again?

**Comment by snowystinger** [comment]
I've found one issue that appears to be broken in this branch but not in main
Go to the currency switcher story
use accounting notation and switch the currency to SAR and the application locale to Arabic
scroll
all scrolling results in large steps in the negative direction by orders of magnitude until it gets stuck

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts (hunk 4 above)
Comment (devongovett) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts:58]: Yeah but we have to compute the cache key still... 🤷

**Review by snowystinger — APPROVED** [review]
LGTM

**Review by dannify — APPROVED** [review]

**Pull request merged**